metals/src/main/scala/scala/meta/internal/metals/MetalsLspService.scala

```diff
@@ -2215,12 +2222,38 @@ class MetalsLspService(
   private def onBuildChangedUnbatched(
       paths: Seq[AbsolutePath]
   ): Future[BuildChange] = {
-    val isBuildChange = paths.exists(buildTools.isBuildRelated(folder, _))
-    if (isBuildChange) {
-      slowConnectToBuildServer(forceImport = false)
-    } else {
-      Future.successful(BuildChange.None)
+    val changedBuilds = paths.flatMap(buildTools.isBuildRelated)
+    val buildChange = for {
+      chosenBuildTool <- tables.buildTool.selectedBuildTool()
+      if (changedBuilds.contains(chosenBuildTool))
+    } yield slowConnectToBuildServer(forceImport = false)
+    buildChange.getOrElse(Future.successful(BuildChange.None))
+  }
+
+  private def onBuildToolAdded(
+      path: AbsolutePath
+  ): Future[BuildChange] = {
+    val supportedBuildTools = buildTools.loadSupported()
+    val maybeBuildChange = for {
+      currentBuildToolName <- tables.buildTool.selectedBuildTool()
+      currentBuildTool <- supportedBuildTools.find(
+        _.executableName == currentBuildToolName
+      )
+      addedBuildName <- buildTools.isBuildRelated(path)
+      if (buildTools.newBuildTool(addedBuildName))
+      if (addedBuildName != currentBuildToolName)
+      newBuildTool <- supportedBuildTools.find(
+        _.executableName == addedBuildName
```

Comment:
On line 2243 we check if this build tool was not already loaded, so 2244 seems redundant? Also, in next line we are searching for `addedBuildName`, but doesn't it run only if the name was not there before?

Reply:
1. `2244` is a sanity check. For a `build tool` to be added to `currentBuildTools` `onBuildToolAdded` needs to be called. It's called by `didSave` and on some file watcher events and I'm not sure it's reliable to always be called properly, thus the sanity check. But maybe it is redundant.

2. Yes, but those are two different things: `val currentBuildTools` are the ones we remember to be currently supported, `def supportedBuildTools` are resolved based on workspace upon call.